clmul/src/backend/autodetect.rs

```diff
@@ -0,0 +1,247 @@
+//! Autodetection for CPU intrinsics, with fallback to the "soft" backend when
+//! they are unavailable.
+
+use core::ops::{BitXor, BitXorAssign};
+
+use crate::backend::soft;
+
+#[cfg(all(target_arch = "aarch64", clmul_armv8))]
+use super::pmull as intrinsics;
+
+#[cfg(any(target_arch = "x86_64", target_arch = "x86"))]
+use super::clmul as intrinsics;
+
+#[cfg(all(target_arch = "aarch64", clmul_armv8))]
+cpufeatures::new!(mul_intrinsics, "aes"); // `aes` implies PMULL
+
+#[cfg(any(target_arch = "x86_64", target_arch = "x86"))]
+cpufeatures::new!(mul_intrinsics, "pclmulqdq");
+
+/// Carryless multiplication
+#[derive(Clone, Copy)]
+pub struct Clmul {
+    inner: Inner,
+    token: mul_intrinsics::InitToken,
+}
+
+#[derive(Clone, Copy)]
+union Inner {
+    intrinsics: intrinsics::Clmul,
+    soft: soft::Clmul,
+}
+
+impl mul_intrinsics::InitToken {
+    #[inline(always)]
+    fn get_intr(&self) -> bool {
+        !cfg!(clmul_force_soft) && self.get()
+    }
+}
+
+impl core::fmt::Debug for Clmul {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        unsafe {
+            if self.token.get_intr() {
+                self.inner.intrinsics.fmt(f)
+            } else {
+                self.inner.soft.fmt(f)
+            }
+        }
+    }
+}
+
+impl Clmul {
+    pub fn new(h: &[u8; 16]) -> Self {
+        let (token, has_intrinsics) = mul_intrinsics::init_get();
+
+        let inner = if cfg!(clmul_force_soft) || !has_intrinsics {
```

Comment:
if we don't have `intrinsics` and we didn't pass `clmul_force_soft`, this code will select `intrinsics` in the else{} branch below. But we need `soft` instead.

Reply:
you are right, sry